Display error type (error, warning, script error) in OS::print_error + cleanup error ANSI coloring

akien-mga opened the pull request:
**tl;dr:** In the end, it looks like this:
- _Dark background colour:_
  ![gt-colour1](https://cloud.githubusercontent.com/assets/4701338/11056936/1560544a-8787-11e5-80b8-b4aa6f7c56ff.jpg)
- _Light background colour:_
  ![gt-colour2](https://cloud.githubusercontent.com/assets/4701338/11056937/17db05da-8787-11e5-8d4c-5ded24b2b4c9.jpg)

---

Previously all types of errors would be shown as ERROR, thus making for example warnings (WARN_PRINT) somewhat aggressive.

Fixes #1127.

Note that I'm a beginner C++ developer, so the coding logic I adopted might not be the best one. I also only tested the changes on Unix/X11, so the changes for Windows and WinRT need to be thoroughly reviewed (though they're more or less copy-pastes of the Unix code).

akien-mga committed:
Display error type (error, warning, script error) in OS::print_error

Previously all types of errors would be shown as ERROR, thus making for example warnings (WARN_PRINT) somewhat aggressive.
ERROR is displayed in red, WARNING in yellow and SCRIPT ERROR in magenta (though the latter does not seem used so far).
Fixes #1127.

akien-mga commented:
I also chose the colours at random:
* ERROR is in red, as previously
* WARNING is in yellow
* SCRIPT ERROR is in green

volzhs commented on core/os/os.cpp:
```diff
@@ -61,9 +61,16 @@ void OS::debug_break() {
 
+	const char* err_type;
+	switch(p_type) {
+		case ERR_ERROR: err_type="**WARNING**"; break;
```
should this ``err_type`` be ``"**ERROR**"``?

akien-mga replied:
Indeed, nice catch. I'll fix that.

akien-mga replied:
This should now be fixed in bb6c3f5.

akien-mga committed:
Fix error messages forcing a white font for subsequent messages

This is achieved using the "no specific formatting" \E[0m tag.
Fixes #2566.

Also remove the hardcoded black background colour and use default bolded terminal font for error message.
Error logs should now look good both on terminals with a dark and light background colour.

akien-mga commented:
I added another commit to reset the terminal colour to the user setting instead of enforcing a white font after an error message.

akien-mga commented:
Rebased the above commit to also remove the hardcoded background colour and use the terminal colour (in bold) instead of a hardcoded white colour for the error log.

In the end, it looks like this:
* Dark background colour:
![gt-colour1](https://cloud.githubusercontent.com/assets/4701338/11056936/1560544a-8787-11e5-80b8-b4aa6f7c56ff.jpg)
* Light background colour:
![gt-colour2](https://cloud.githubusercontent.com/assets/4701338/11056937/17db05da-8787-11e5-8d4c-5ded24b2b4c9.jpg)

bojidar-bg commented:
:+1: Looks nice :smile:

ZuBsPaCe commented:
Looks good on windows! (compiled with vs2015)

![godot-error-types-windows](https://cloud.githubusercontent.com/assets/4233942/11065631/bf5494ec-87c4-11e5-8cd5-f76d829b82bb.png)

Some things I noticed:
* I'm not able to generate SCRIPT ERRORS without calling _err_print_error explicitly with that type. How do you generate those? As you can see in the screenshot, at the end I'm calling a method on a null-instance. Shouldn't this be a script error?
* In the screenshot I called print and printerr. In my opinion those should be colored, too. Maybe green and red?
* Green stands for 'ok', which does not work well with SCRIPT_ERROR... What about using purple or maybe orange?

akien-mga commented:
> I'm not able to generate SCRIPT ERRORS without calling _err_print_error explicitly with that type. How do you generate those? As you can see in the screenshot, at the end I'm calling a method on a null-instance. Shouldn't this be a script error?

Yeah I added a formatting for script errors as they are part of the ErrorType enum, but I'm not sure they are really used right now. I would indeed expect your error in _ready() to be a script error, if @reduz can confirm maybe you could open a new issue to enable the use of script errors throughout the code?

> In the screenshot I called print and printerr. In my opinion those should be colored, too. Maybe green and red?

I don't think print should be coloured, but maybe printerr could be in red yes.

> Green stands for 'ok', which does not work well with SCRIPT_ERROR... What about using purple or maybe orange?

Ah yes that's a good point. I'll try purple or orange then.

akien-mga commented:
@ZuBsPaCe The available colours are actually pretty limited: https://en.wikipedia.org/wiki/ANSI_escape_code#Colors
I played around with some values, see e.g. here cyan and magenta:
![gt-colour3](https://cloud.githubusercontent.com/assets/4701338/11069778/38e69374-87d8-11e5-878a-778eba8546bf.jpg)

It's not perfect, but I guess I'll use magenta for the script error. Blue is also cool but the dark blue is way too dark to be readable on a black background.

akien-mga commented:
I rebase to have a cleaner PR, the first commit does most of what was described initially plus uses magenta for script error instead of green.
The second commit adds some cosmetic improvements by removing some hardcoded font and background colours.

vnen commented:
Looks good on Windows, yes, unless you set custom colors. The errors always use black background. 

![windows-godot-console-color](https://cloud.githubusercontent.com/assets/5599796/11078172/5df26d7e-87eb-11e5-97c7-6e37c4ee3c32.png)

I believe the Windows color coding is happening afterwards, since the prompt does not support ANSI codes. Check [these lines](https://github.com/godotengine/godot/blob/61c4c5795ff9057b1c67142758418a4f982a0241/platform/windows/os_windows.cpp#L1781-L1832).

akien-mga commented:
Thanks for the test @vnen. I don't have experience with Windows, could you maybe see if you can find the parameters to make the output similar to that of ANSI-compatible terminals like in the screenshots in https://github.com/godotengine/godot/pull/2737#issuecomment-155347757?

akien-mga commented:
Note that I did not change the colouring scheme for non ANSI-compatible Windows terminals though, so this is not a regression.

ZuBsPaCe commented:
@akien-mga I like purple for script errors.

About the custom colors mentioned by @vnen. It's possible to determine the background color which can be used in the subsequent SetConsoleTextAttribute calls. So after basecol is set you do:

    uint32_t backcol = sbi.wAttributes & ~(FOREGROUND_RED|FOREGROUND_GREEN|FOREGROUND_BLUE|FOREGROUND_INTENSITY);

and then you need to add backcol to all SetConsoleTextAttribute calls except the last one, which resets the attributes, for example:

    SetConsoleTextAttribute(hCon,basecol|FOREGROUND_INTENSITY|backcol); 

Here's the result:

![godot-error-types-windows-white](https://cloud.githubusercontent.com/assets/4233942/11080080/32b8c3a8-8811-11e5-9d4f-0e9ad88eacb4.png)

vnen commented:
I was able to fix this. The following diff upon your current commits is working fine for me. Also remove escape sequences from windows, since it doesn't work and `\E` gives compilation warnings as non-recognized escape sequence (and put a missing `\n` in script error):

```diff
diff --git a/platform/windows/os_windows.cpp b/platform/windows/os_windows.cpp
index 964b9e7..579ddd0 100644
--- a/platform/windows/os_windows.cpp
+++ b/platform/windows/os_windows.cpp
@@ -1777,16 +1777,16 @@ void OS_Windows::print_error(const char* p_function,const char* p_file,int p_lin
 
 		switch(p_type) {
 			case ERR_ERROR:
-				print("\E[1;31mERROR: %s: \E[0m\E[1m%s\n",p_function,err_details);
-				print("\E[0;31m   At: %s:%i.\E[0m\n",p_file,p_line);
+				print("ERROR: %s: %s\n",p_function,err_details);
+				print("   At: %s:%i.\n",p_file,p_line);
 				break;
 			case ERR_WARNING:
-				print("\E[1;33mWARNING: %s: \E[0m\E[1m%s\n",p_function,err_details);
-				print("\E[0;33m     At: %s:%i.\E[0m\n",p_file,p_line);
+				print("WARNING: %s: %s\n",p_function,err_details);
+				print("     At: %s:%i.\n",p_file,p_line);
 				break;
 			case ERR_SCRIPT:
-				print("\E[1;35mSCRIPT ERROR: %s: \E[0m\E[1m",p_function,err_details);
-				print("\E[0;35m          At: %s:%i.\E[0m\n",p_file,p_line);
+				print("SCRIPT ERROR: %s: %s\n",p_function,err_details);
+				print("          At: %s:%i.\n",p_file,p_line);
 				break;
 		}
 	} else {
@@ -1794,17 +1794,18 @@ void OS_Windows::print_error(const char* p_function,const char* p_file,int p_lin
 		CONSOLE_SCREEN_BUFFER_INFO sbi; //original
 		GetConsoleScreenBufferInfo(hCon,&sbi);
 
-		SetConsoleTextAttribute(hCon,sbi.wAttributes);
-
-
+		WORD current_bg = sbi.wAttributes & (BACKGROUND_RED|BACKGROUND_GREEN|BACKGROUND_BLUE|BACKGROUND_INTENSITY);
+		WORD current_fg = sbi.wAttributes & (FOREGROUND_RED|FOREGROUND_GREEN|FOREGROUND_BLUE|FOREGROUND_INTENSITY);
 
 		uint32_t basecol=0;
 		switch(p_type) {
 			case ERR_ERROR: basecol = FOREGROUND_RED; break;
 			case ERR_WARNING: basecol = FOREGROUND_RED|FOREGROUND_GREEN; break;
-			case ERR_SCRIPT: basecol = FOREGROUND_GREEN; break;
+			case ERR_SCRIPT: basecol = FOREGROUND_RED|FOREGROUND_BLUE; break;
 		}
 
+		basecol |= current_bg;
+
 		if (p_rationale && p_rationale[0]) {
 
 			SetConsoleTextAttribute(hCon,basecol|FOREGROUND_INTENSITY);
@@ -1816,11 +1817,11 @@ void OS_Windows::print_error(const char* p_function,const char* p_file,int p_lin
 				case ERR_SCRIPT: print("SCRIPT ERROR: "); break;
 			}
 
-			SetConsoleTextAttribute(hCon,FOREGROUND_RED|FOREGROUND_BLUE|FOREGROUND_GREEN|FOREGROUND_INTENSITY);
+			SetConsoleTextAttribute(hCon,current_fg|current_bg);
 			print(" %s\n",p_rationale);
 			SetConsoleTextAttribute(hCon,basecol);
 			print("At: ");
-			SetConsoleTextAttribute(hCon,FOREGROUND_RED|FOREGROUND_BLUE|FOREGROUND_GREEN);
+			SetConsoleTextAttribute(hCon,current_fg|current_bg);
 			print(" %s:%i\n",p_file,p_line);
 
 
@@ -1831,11 +1832,11 @@ void OS_Windows::print_error(const char* p_function,const char* p_file,int p_lin
 				case ERR_WARNING: print("WARNING: %s: ",p_function); break;
 				case ERR_SCRIPT: print("SCRIPT ERROR: %s: ",p_function); break;
 			}
-			SetConsoleTextAttribute(hCon,FOREGROUND_RED|FOREGROUND_BLUE|FOREGROUND_GREEN|FOREGROUND_INTENSITY);
+			SetConsoleTextAttribute(hCon,current_fg|current_bg);
 			print(" %s\n",p_code);
 			SetConsoleTextAttribute(hCon,basecol);
 			print("At: ");
-			SetConsoleTextAttribute(hCon,FOREGROUND_RED|FOREGROUND_BLUE|FOREGROUND_GREEN);
+			SetConsoleTextAttribute(hCon,current_fg|current_bg);
 			print(" %s:%i\n",p_file,p_line);
 		}
```

As per WinRT, I don't think it has any console. Might be good to remove escape sequences there too, since it'll likely complain about `\E`.

Finally, might be needing some spaces before "At:" to be like the *nix version.
